src/libraries/System.IO.FileSystem/tests/FileStream/SafeFileHandle.cs

```diff
@@ -104,14 +104,7 @@ private async Task ThrowWhenHandlePositionIsChanged(bool useAsync)
 
                     fs.WriteByte(0);
                     fsr.Position++;
-                    if (useAsync && OperatingSystem.IsWindows()) // Async I/O behaviors differ due to kernel-based implementation on Windows
-                    {
-                        Assert.Throws<IOException>(() => FSAssert.CompletesSynchronously(fs.ReadAsync(new byte[1], 0, 1)));
-                    }
-                    else
-                    {
-                        await Assert.ThrowsAsync<IOException>(() => fs.ReadAsync(new byte[1], 0, 1));
-                    }
+                    await Assert.ThrowsAsync<IOException>(() => fs.ReadAsync(new byte[1], 0, 1));
```

Comment:
It still throws exactly the same exception but does not complete synchronously. It's caused by the fact that `BufferedStream` serializes the async reads and writes.

Reply:
> It's caused by the fact that BufferedStream serializes the async reads and writes.

Ok, but what other operation is already in flight that would cause this to now be delayed?

Reply:
@stephentoub after thinking about it for a second I decided to revert my change, add the legacy condition and a comment that hopefully answers your initial question